// @loaders.gl/tables/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TableBatchBuilder", {
  enumerable: true,
  get: function get() {
    return _tableBatchBuilder["default"];
  }
});
Object.defineProperty(exports, "RowTableBatch", {
  enumerable: true,
  get: function get() {
    return _rowTableBatch["default"];
  }
});
Object.defineProperty(exports, "ColumnarTableBatch", {
  enumerable: true,
  get: function get() {
    return _columnarTableBatch["default"];
  }
});
Object.defineProperty(exports, "deduceTableSchema", {
  enumerable: true,
  get: function get() {
    return _schemaUtils.deduceTableSchema;
  }
});
Object.defineProperty(exports, "JSONLoader", {
  enumerable: true,
  get: function get() {
    return _jsonLoader["default"];
  }
});
Object.defineProperty(exports, "XMLLoader", {
  enumerable: true,
  get: function get() {
    return _xmlLoader["default"];
  }
});
Object.defineProperty(exports, "AsyncQueue", {
  enumerable: true,
  get: function get() {
    return _asyncQueue["default"];
  }
});

var _tableBatchBuilder = _interopRequireDefault(require("./lib/table/table-batch-builder"));

var _rowTableBatch = _interopRequireDefault(require("./lib/table/row-table-batch"));

var _columnarTableBatch = _interopRequireDefault(require("./lib/table/columnar-table-batch"));

var _schemaUtils = require("./lib/table/schema-utils");

var _jsonLoader = _interopRequireDefault(require("./json-loader"));

var _xmlLoader = _interopRequireDefault(require("./xml-loader"));

var _asyncQueue = _interopRequireDefault(require("./lib/utils/async-queue"));
