// @loaders.gl/tables/dist/esm/lib/table/table-batch-builder.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
var DEFAULT_BATCH_SIZE = 100;

var TableBatchBuilder = function () {
  function TableBatchBuilder(TableBatchType, schema) {
    var batchSize = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : DEFAULT_BATCH_SIZE;

    _classCallCheck(this, TableBatchBuilder);

    this.TableBatchType = TableBatchType;
    this.schema = schema;
    this.batchSize = batchSize;
    this.batch = null;
    this.batchCount = 0;
  }

  _createClass(TableBatchBuilder, [{
    key: "addRow",
    value: function addRow(row) {
      if (!this.batch) {
        var TableBatchType = this.TableBatchType;
        this.batch = new TableBatchType(this.schema, this.batchSize);
      }

      this.batch.addRow(row);
    }
  }, {
    key: "chunkComplete",
    value: function chunkComplete() {
      if (this.batch) {
        this.batch.chunkComplete();
      }
    }
  }, {
    key: "isFull",
    value: function isFull() {
      return this.batch && this.batch.isFull();
    }
  }, {
    key: "hasBatch",
    value: function hasBatch() {
      return Boolean(this.batch);
    }
  }, {
    key: "getNormalizedBatch",
    value: function getNormalizedBatch() {
      if (this.batch) {
        var normalizedBatch = this.batch.getNormalizedBatch();
        this.batch = null;
        normalizedBatch.count = this.batchCount;
        this.batchCount++;
        return normalizedBatch;
      }

      return null;
    }
  }]);

  return TableBatchBuilder;
}();

export { TableBatchBuilder as default };
