// @loaders.gl/zip/dist/es5/zip-writer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZipWriter = void 0;

var _jszip = _interopRequireDefault(require("jszip"));

var ZipWriter = {
  name: 'Zip Archive',
  extensions: ['zip'],
  category: 'archive',
  mimeType: 'application/zip',
  encode: encodeZipAsync
};
exports.ZipWriter = ZipWriter;

function encodeZipAsync(fileMap, options) {
  var jsZip = new _jszip["default"]();

  for (var subFileName in fileMap) {
    var subFileData = fileMap[subFileName];
    jsZip.file(subFileName, subFileData, options);
  }

  options = Object.assign({}, options, {
    type: 'arraybuffer'
  });
  var _options = options,
      _options$onUpdate = _options.onUpdate,
      onUpdate = _options$onUpdate === void 0 ? function () {} : _options$onUpdate;
  return jsZip.generateAsync(options, onUpdate)["catch"](function (error) {
    options.log.error("Unable to write zip archive: ".concat(error));
    throw error;
  });
}
